### PR TITLE
MM-29726 Allow disabling link previews from certain domains

### PR DESCRIPTION
**Pull request by berkeka**
#### Summary

This pull-request implements the feature where an admin can restrict certain links from creating a preview. This can be done by adding the domain to the restrict link preview field in System Console > Configuration > Posts.

#### Ticket Link

https://github.com/mattermost/mattermost-server/issues/16846

#### Release Note
```release-note
* Added a string field to config for restricted domains with the key RestrictLinkPreviews 
* Added a UI field for restricted domains under Site Configuration > Posts in the System Console
* Expanded the logic that determines whether a post has a preview or not
```

### CODE REVIEW

**Review by hmhealey — APPROVED** [review]
Thanks for the pull request, @berkeka! This looks good to me. 

Good idea to check the link in `getFirstLinkAndImages`. I was actually thinking we'd do it in `getLinkMetadata`, but that has the downside of not letting us fall through to the next link in a post if a user were to post multiple. I know that would've come up in the use case we want to address for ourselves since we wanted to block previews generated by a glossary plugin that we use. This does require us to be a bit more careful with how we use `getLinkMetadata`, but that should be fine.

File: app/post_metadata.go [hunk 1]
```diff
@@ -306,23 +306,44 @@ func (a *App) getCustomEmojisForPost(post *model.Post, reactions []*model.Reacti
...
+	domains := a.normalizeDomains(*a.Config().ServiceSettings.RestrictLinkPreviews)
+	matched := false
+	for _, d := range domains {
+		if strings.Contains(link, d) {
+			matched = true
+			break
+		}
+	}
+
+	return !matched
```
Comment (isacikgoz) [inline app/post_metadata.go:319]:
```suggestion
	for _, d := range domains {
		if strings.Contains(link, d) {
			return false
		}
	}

	return true
```

File: app/post_metadata.go [hunk 2]
```diff
@@ -306,23 +306,44 @@ func (a *App) getCustomEmojisForPost(post *model.Post, reactions []*model.Reacti
...
 
 	markdown.Inspect(str, func(blockOrInline interface{}) bool {
+		link := ""
 		switch v := blockOrInline.(type) {
 		case *markdown.Autolink:
-			if firstLink == "" {
-				firstLink = v.Destination()
+			link = v.Destination()
+			if firstLink == "" && a.isLinkAllowedForPreview(link) {
+				firstLink = link
```
Comment (isacikgoz) [inline app/post_metadata.go:335]: we can use shorthand `if` in every case, like: 
```suggestion
			
			if link := v.Destination(); firstLink == "" && a.isLinkAllowedForPreview(link) {
				firstLink = link
```

File: app/post_metadata.go [hunk 3]
```diff
@@ -306,23 +306,44 @@ func (a *App) getCustomEmojisForPost(post *model.Post, reactions []*model.Reacti
...
 // Given a string, returns the first autolinked URL in the string as well as an array of all Markdown
 // images of the form ![alt text](image url). Note that this does not return Markdown links of the
 // form [text](url).
-func getFirstLinkAndImages(str string) (string, []string) {
+func (a *App) getFirstLinkAndImages(str string) (string, []string) {
 	firstLink := ""
 	images := []string{}
 
 	markdown.Inspect(str, func(blockOrInline interface{}) bool {
+		link := ""
```
Comment (isacikgoz) [inline app/post_metadata.go:330]:
```suggestion
```

File: app/post_metadata_test.go [hunk 4]
```diff
@@ -1304,7 +1307,67 @@ func TestGetFirstLinkAndImages(t *testing.T) {
...
+[logo]: http://example.com/logo`,
+			ExpectedFirstLink: "",
+			ExpectedImages:    []string{},
+		},
+		"image and link domain is restricted": {
+			Input:             "this is a https://example.com and ![our logo](https://example.com/logo)",
+			ExpectedFirstLink: "",
+			ExpectedImages:    []string{},
+		},
+	} {
```
Comment (isacikgoz) [inline app/post_metadata_test.go:1364]: Can we add a case where we expect an image?

File: model/config.go [hunk 5]
```diff
@@ -296,6 +296,7 @@ type ServiceSettings struct {
 	EnablePostUsernameOverride                        *bool    `access:"integrations"`
 	EnablePostIconOverride                            *bool    `access:"integrations"`
 	EnableLinkPreviews                                *bool    `access:"site"`
+	RestrictLinkPreviews                              *string  `access:"site"`
```
Comment (isacikgoz) [inline model/config.go:299]: @wiersgallak shall we add this new setting to the telemetries?

**Review by isacikgoz — CHANGES_REQUESTED** [review]
Awesome work @berkeka 🎉 There are some opportunities to slim down the code, otherwise looks good. I requested additional tests to be more confident about this change.

File: app/post_metadata_test.go (hunk 4 above)
Comment (berkeka) [reply app/post_metadata_test.go:1374]: @isacikgoz Will the newly added cases be enough? d923a53

Comment (isacikgoz) [reply app/post_metadata_test.go:1374]: Yep, thanks!

**Comment by hmhealey** [comment]
@berkeka The build is failing right now since we have no telemetry for whether or not the new setting is used. If you look at `services/telemetry.go`, you'll see where we track the values of different settings and that they're grouped by their settings category. Since this setting allows arbitrary values, we just want to track whether or not it's the default empty value or not.

File: model/config.go (hunk 5 above)
Comment (wiersgallak) [reply model/config.go:303]: @isacikgoz Yes please

**Comment by wiersgallak** [comment]
added @esethna to review the effect to the end-user experience

**Review by esethna — APPROVED** [review]
Reviewed on the client PR

**Review by isacikgoz — APPROVED** [review]
LGTM 👍

**Comment by hmhealey** [comment]
The test server is still using the master branch of the web app. I'm going to try restarting it after both PRs pass all the CI checks to see if that lets it pick up the correct version.

**Comment by hmhealey** [comment]
Discussed with Jelena and Jesse, and we're just going to merge this without QA review so that we can test the web app in that PR because we can't get the test servers to line up properly.

**Pull request merged**